Add firmware install service doc

joelrebel opened the pull request:
Proposal for the firmware install service and related components.

rendered markdown link https://github.com/metal-toolbox/architecture/blob/firmware-install-service/firmware-install-service.md

joelrebel committed:
Add firmware install service doc

mmlb commented on firmware-install-service.md:
```diff
@@ -0,0 +1,334 @@
+{
+ "kind": "firmwareInstall",
+ "status": "pending", // updated by the controller
+ "data": "{ arbitrary JSON by choice of the controller }", // updated by the controller
```
whats the point of json-as-a-string if its specified to be json? Why not just an opaque object `"data": {...}`?

joelrebel replied:
oh right, that is basically what I mean't to have there. fixed [733a45e](https://github.com/metal-toolbox/architecture/pull/1/commits/733a45e992ccd0b89c8b1cd598e8418ba306f07f)

joelrebel committed:
fix some grammar

mmlb commented on firmware-install-service.md:
```diff
@@ -0,0 +1,334 @@
+- Controllers transition and keep the *condition* updated based on its work.
+- Controller send a message on the NATs message bus to update the `state`, `data` attributes.
+- Controllers should not modify the condition `kind` attribute.
+- When updating the `status` or `data` attributes, the controller must pass the original `tally` value from the previous operation where it received this value, Serverservice will not accept a `PUT` update on a Conditions if the `tally` value does not match the existing, this mechanism functions as an [optimistic lock](https://en.wikipedia.org/wiki/Optimistic_concurrency_control).
```
why not just call it `lock` and avoid confusion?

joelrebel replied:
updated

mmlb commented on firmware-install-service.md:
```diff
@@ -0,0 +1,333 @@
+
+#### Locks
+
+- Locks are a Serverservice construct to restrict the number of controllers acting on the same device at the same instant.
```
how is this different than `tally`?

joelrebel replied:
[733a45e](https://github.com/metal-toolbox/architecture/pull/1/commits/733a45e992ccd0b89c8b1cd598e8418ba306f07f) addresses this question. The server `locks` are mainly to prevent multiple controllers acting on the same server BMC at the same instant.

mmlb replied:
oh so the `Locks` documented here is referring to what was earlier named `tally`?

joelrebel replied:
Sorry, since they now share the same name it gets a bit confusing. 

The `lock` (previously `tally`) field in the conditions object is to ensure the data in the condition object is not over written accidentally - out of sequence (let me know if I should elaborate here).

The `locks` construct discussed in this section is a feature to be added to serverservice as another object which is associated with a server, to hold exclusive access to the server BMC (for cases where its necessary).

Hope that clears it up.

joelrebel committed:
Update doc based on feedback

mmlb commented:
fyi I updated the rendered link to a non-commit ref so that its always up2date

diogomatsubara commented:
Thanks for working on this Joel, makes it much easier to visualise how things are going to work. I specially liked that you make Dustin start the party 🎉 

In the "Successful Out of band firmware install" sequence diagram, between steps 8 and 9, do we want to query the server for the firmware install or look at the server inventory for what's installed? That is, do we want to compare the fw-set against what we think is in the server (the inventory and supposedly the source of truth) or we want look in the server and ignore the inventory?

At what point the inventory is updated again with the new information? Correct me if I'm wrong, but I'm assuming the power off will also be a deprovision? If yes, then alloy will collect the new inventory during that phase? What if it's a `FirmwareInstall` that doesn't run deprov, when the inventory is updated?

A nit, the diagram mentions "Artifacts bucket", I think it could be Artifacts https endpoint or just Artifacts, just to make sure we're not taking things directly from the bucket but using the artifacts endpoint.

joelrebel committed:
Update Artifacts node based on feedback

joelrebel commented:
@diogomatsubara thanks for taking a look,

> In the "Successful Out of band firmware install" sequence diagram, between steps 8 and 9, do we want to query the server for the firmware install or look at the server inventory for what's installed? That is, do we want to compare the fw-set against what we think is in the server (the inventory and supposedly the source of truth) or we want look in the server and ignore the inventory?

Theres a few reasons to query the device before installing the firmware, all of these are related to a situation where
the data in the inventory is out of sync with what is actually on the device.

 - If the inventory collection for the device fails.
 - If the installed firmware version matches the desired version, flasher can then skip that firmware install. 
 - If the inventory for the server has not been collected as yet, flasher can perform an install regardless.

Theres also the option that, once the server conditions are functional, the `inventoryOOB` condition could always be run before and after the `firmwareUpdate` condition and then the firmware install can depend on the inventory being the source of truth.

> At what point the inventory is updated again with the new information? Correct me if I'm wrong, but I'm assuming the power off will also be a deprovision? If yes, then alloy will collect the new inventory during that phase? What if it's a FirmwareInstall that doesn't run deprov, when the inventory is updated?

The power off is a plain host power off (only if the host was powered off to begin with) and does not involve any deprovision process. 
Like mentioned above, collecting the inventory after a firmware install could be orchestrated by the `condition-orchestrator` by it setting the `inventoryOOB` condition on the server after it figures that the `firmwareInstall` condition was finalized. This can be done at a later point though, once we're clear on the controller behaviour.

> A nit, the diagram mentions "Artifacts bucket", I think it could be Artifacts https endpoint or just Artifacts, just to make sure we're not taking things directly from the bucket but using the artifacts endpoint.

Thanks, fixed in [623f270](https://github.com/metal-toolbox/architecture/pull/1/commits/623f27085b89fc153e34d5ceb3484c87368e4da0)

diogomatsubara commented:
> @diogomatsubara thanks for taking a look,
> 
> > In the "Successful Out of band firmware install" sequence diagram, between steps 8 and 9, do we want to query the server for the firmware install or look at the server inventory for what's installed? That is, do we want to compare the fw-set against what we think is in the server (the inventory and supposedly the source of truth) or we want look in the server and ignore the inventory?
> 
> Theres a few reasons to query the device before installing the firmware, all of these are related to a situation where the data in the inventory is out of sync with what is actually on the device.

If the inventory is out of sync, server service is not a very good source of truth.

> 
> * If the inventory collection for the device fails.
> * If the installed firmware version matches the desired version, flasher can then skip that firmware install.
> * If the inventory for the server has not been collected as yet, flasher can perform an install regardless.
> 
> Theres also the option that, once the server conditions are functional, the `inventoryOOB` condition could always be run before and after the `firmwareUpdate` condition and then the firmware install can depend on the inventory being the source of truth.
> 

I think this is a good option.

> > At what point the inventory is updated again with the new information? Correct me if I'm wrong, but I'm assuming the power off will also be a deprovision? If yes, then alloy will collect the new inventory during that phase? What if it's a FirmwareInstall that doesn't run deprov, when the inventory is updated?
> 
> The power off is a plain host power off (only if the host was powered off to begin with) and does not involve any deprovision process. Like mentioned above, collecting the inventory after a firmware install could be orchestrated by the `condition-orchestrator` by it setting the `inventoryOOB` condition on the server after it figures that the `firmwareInstall` condition was finalized. This can be done at a later point though, once we're clear on the controller behaviour.
> 

Sounds good to me, thanks for the clarifications.

ofaurax commented:
Thanks for the doc!

I see that the flasher downloads/verifies/installs the BMC firmware, then downloads/verifis/installs BIOS firmware.
Shouldn't we download/verify all needed firmwares before flashing all?

This would ensure that all components are updated together, but might also prevent a BMC update if BIOS firmware is corrupted or unavailable.

joelrebel committed:
firmware-install-service: update a few more bits based on feedback

joelrebel commented:
> I see that the flasher downloads/verifies/installs the BMC firmware, then downloads/verifis/installs BIOS firmware.
Shouldn't we download/verify all needed firmwares before flashing all?

> This would ensure that all components are updated together, but might also prevent a BMC update if BIOS firmware is corrupted or unavailable.

Thats a good point, I've updated the doc and added an issue to follow up on. Thanks!

joelrebel committed:
firmware-install-service: update diagrams and notes based on feedback

joelrebel commented:
most of the questions and suggestions here were addressed, going ahead and merging this for now.
